tensorflow_quantum/datasets/spin_system.py

```diff
@@ -312,6 +312,248 @@ def tfi_chain(qubits, boundary_condition="closed", data_dir=None):
+        data_dir: Optional Python `str` location where to store the data on
+            disk. Defaults to `/tmp/.keras`.
+    Returns:
+        A Python `lst` cirq.Circuit of depth len(qubits) / 2 with resolved
+            parameters.
+        A Python `lst` of labels, 0, for the critical metallic phase
+            (`Delta<=1`) and 1 for the insulating phase (`Delta>1`).
+        A Python `lst` of `cirq.PauliSum`s.
+        A Python `lst` of `namedtuple` instances containing the following
+            fields:
+        - `g`: Numpy `float` order parameter.
+        - `gs`: Complex `np.ndarray` ground state wave function from
+            exact diagonalization.
+        - `gs_energy`: Numpy `float` ground state energy from exact
+            diagonalization.
+        - `res_energy`: Python `float` residual between the circuit energy
+            and the exact energy from exact diagonalization.
+        - `fidelity`: Python `float` overlap between the circuit state
+            and the exact ground state from exact diagonalization.
+        - `params`: Dict with Python `str` keys and Numpy`float` values.
+            Contains $M \times P $ parameters. Here $M$ is the number of
+            parameters per circuit layer and $P$ the circuit depth.
+        - `var_circuit`: Variational `cirq.Circuit` quantum circuit with
+            unresolved Sympy parameters.
+    """
+
+    supported_n = [4, 8, 12, 16]
+    supported_bc = ["closed"]
+    if any(isinstance(q, list) for q in qubits):
+        raise TypeError("qubits must be a one-dimensional list")
+
+    if not all(isinstance(q, cirq.GridQubit) for q in qubits):
+        raise TypeError("qubits must be a list of cirq.GridQubit objects.")
+
+    nspins = len(qubits)
+    depth = nspins // 2
+    if nspins not in supported_n:
+        raise ValueError("Supported number of spins are {}, received {}".format(
+            supported_n, nspins))
+
+    if boundary_condition not in supported_bc:
+        raise ValueError(
+            "Supported boundary conditions are {}, received {}".format(
+                supported_bc, boundary_condition))
+
+    data_path = _download_spin_data('XXZ_chain', boundary_condition, nspins,
+                                    data_dir)
+
+    name_generator = unique_name()
+
+    # 4 * N/2 parameters.
+    symbol_names = [next(name_generator) for _ in range(2 * nspins)]
+    symbols = [sympy.Symbol(name) for name in symbol_names]
+
+    # Define the circuit.
+    circuit = cirq.Circuit(cirq.X.on_each(qubits))
+    circuit.append(cirq.H(qubits[i]) for i in range(0, nspins, 2))
+    circuit.append(
+        cirq.CNOT(qubits[j], qubits[j + 1]) for j in range(0, nspins - 1, 2))
+
+    odd_qubits = list(
+        zip([qubits[j] for j in range(1, nspins - 1, 2)],
+            [qubits[j + 1] for j in range(1, nspins - 1, 2)]))
+    even_qubits = list(
+        zip([qubits[j] for j in range(0, nspins - 1, 2)],
+            [qubits[j + 1] for j in range(0, nspins - 1, 2)]))
```

Comment:
The interchange of zip statements and slightly different indices is kind of confusing here. Might make a lot more sense if when defining the circuit we did something like:

```python3
even_qubits = qubits[::2]
odd_qubits = qubits[1::2]

circuit.append(cirq.H.on_each(even_qubits))
circuit.append(cirq.CNOT(e, o) for e, o in zip(even_qubits, odd_qubits))
.
.
.

```